### PR TITLE
fix(curriculum): refactor workshop cafe menu assertions to use DOM

### LINKED ISSUE #69075: Cafe Menu workshop hints test source text instead of the rendered DOM
### Describe the Issue

Several hints in the `Design a Cafe Menu` workshop check the learner's raw source with a regex on `code` for something the rendered document already answers directly.

1. Several hints match the learner's source text for something that is directly observable in the rendered document. Element presence:

```js
assert.match(code, /<main>/i);
```

Steps 1, 2, 3, 4, 5, 25, 59, and 63 do this for `main`, `h1`, `p`, `section`, `h2`, `article`, `footer`, and `hr`.

2. Step 6 checks nesting with a source regex when a descendant selector expresses it.

```js
assert.match(code, /<head\s*>[\w\W\s]*<style\s*>[\w\W\s]*<\/style\s*>[\w\W\s]*<\/head\s*>/i)
```

3. Step 13 counts elements and reads attributes by matching source text.

```js
assert.lengthOf(code.match(/<meta.*?\/?>/g),2);
```

```js
assert.exists(meta[0]?.outerHTML?.match(/name=('|")viewport\1/) || meta[1]?.outerHTML?.match(/name=('|")viewport\1/));
```

4. Step 42 counts sections the same way.

```js
assert.lengthOf(code.match(/<section>/ig) ,2);
```

5. Steps 57, 73, and 85 match a `class` attribute in the source when the class is queryable, and in each case the next hint in the same file already queries it.

```js
assert.match(code,/<p class=('|")established\1>/i);
```

6. Steps 36, 39, 46, and 48 check for whitespace between two `p` elements by matching source text.

```js
assert.match(code, /Vanilla<\/p><p/);
```

```js
assert.notMatch(code, /<\/p>\s+<p/);
```

The condition is visible in the parsed document. `<p>a</p>\n  <p>b</p>` parses to an element, a text node holding the newline and indentation, then the second element, so the first paragraph's `nextSibling` is that text node. `<p>a</p><p>b</p>` parses with nothing between them, so `nextSibling` is the second element. Asserting on that reads the same condition from the DOM:

```js
const flavor = document.querySelector('.flavor');
const price = document.querySelector('.price');
assert.strictEqual(flavor?.nextSibling, price);
```

For the steps that cover every menu item:

```js
document.querySelectorAll('.item').forEach(item => {
  assert.lengthOf(item.children, 2);
  const [first, second] = item.children;
  assert.strictEqual(first.nextSibling, second);
});
```

The child-count assertion matters because a learner who deleted one paragraph would otherwise get `strictEqual(node, undefined)`, which does not describe what went wrong.

The regex also has a wider reach than the step it belongs to: it forbids the pattern anywhere in the file, so it constrains paragraphs the step never mentions. Carried past Step 62 it would fail on the footer, where `</p>` and `<p class="address">` legitimately sit on separate lines.

7. Step 11's first hint reads `You should not have any style tags in your code`, but the assertion tests for the bare substring `style` across both files rather than for a `style` element. A learner who adds `<link href="styles.css" rel="stylesheet">` ahead of Step 12 trips it and is told they still have style tags. The hint directly below it already tests the actual requirement against the editable region.

```js
assert.notMatch(code, /style/i);
```

### Affected Pages

- https://www.freecodecamp.org/learn/responsive-web-design-v9/workshop-cafe-menu/step-1
- https://www.freecodecamp.org/learn/responsive-web-design-v9/workshop-cafe-menu/step-2
- https://www.freecodecamp.org/learn/responsive-web-design-v9/workshop-cafe-menu/step-3
- https://www.freecodecamp.org/learn/responsive-web-design-v9/workshop-cafe-menu/step-4
- https://www.freecodecamp.org/learn/responsive-web-design-v9/workshop-cafe-menu/step-5
- https://www.freecodecamp.org/learn/responsive-web-design-v9/workshop-cafe-menu/step-6
- https://www.freecodecamp.org/learn/responsive-web-design-v9/workshop-cafe-menu/step-11
- https://www.freecodecamp.org/learn/responsive-web-design-v9/workshop-cafe-menu/step-13
- https://www.freecodecamp.org/learn/responsive-web-design-v9/workshop-cafe-menu/step-25
- https://www.freecodecamp.org/learn/responsive-web-design-v9/workshop-cafe-menu/step-36
- https://www.freecodecamp.org/learn/responsive-web-design-v9/workshop-cafe-menu/step-39
- https://www.freecodecamp.org/learn/responsive-web-design-v9/workshop-cafe-menu/step-42
- https://www.freecodecamp.org/learn/responsive-web-design-v9/workshop-cafe-menu/step-46
- https://www.freecodecamp.org/learn/responsive-web-design-v9/workshop-cafe-menu/step-48
- https://www.freecodecamp.org/learn/responsive-web-design-v9/workshop-cafe-menu/step-57
- https://www.freecodecamp.org/learn/responsive-web-design-v9/workshop-cafe-menu/step-59
- https://www.freecodecamp.org/learn/responsive-web-design-v9/workshop-cafe-menu/step-63
- https://www.freecodecamp.org/learn/responsive-web-design-v9/workshop-cafe-menu/step-73
- https://www.freecodecamp.org/learn/responsive-web-design-v9/workshop-cafe-menu/step-85

### Expected behavior

1. Checks for element presence, nesting, counts, attribute values, and class membership run against the rendered document rather than the learner's source text, so reformatting valid HTML does not fail a step. Regex on `code` remains only where nothing in the DOM expresses the condition, such as the absence of a closing tag on a void element or the presence of a CSS comment.

   Selectors on `head`, `meta`, `link`, and `title` behave normally in the test environment; `lab-recipe-page`, `lab-travel-agency-page`, `lab-business-card`, and `workshop-accessibility-quiz` all query them directly. The `style` element is the exception: the environment injects one of its own, so a count or presence check on `style` needs to exclude it, as `basic-css` does with `document.querySelectorAll('style:not(.fcc-hide-header)')` and as Step 8 of this workshop does with `assert.isAtMost(document.querySelectorAll('style').length, 2)`.

2. The whitespace checks in Steps 36, 39, 46, and 48 assert that the two paragraphs of an item are adjacent siblings in the parsed document, and they cover only the paragraphs the step is about rather than every paragraph on the page.

3. Step 11's first hint tests for the presence of a `style` element, matching what its text says, so a link to `styles.css` does not trip it.

### PR DESCRIPTION
**Pull request by rohit-h11**
Checklist:

- [x] I have read and followed the contribution guidelines.
- [x] I have read and followed the how to open a pull request guide.
- [x] My pull request targets the main branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitHub Codespaces.

Closes #69075

### Description
This PR refactors fragile, regex-based source-code checking tests (`assert.match(code, /.../)/`) in the **Design a Cafe Menu (workshop)** curriculum to use robust DOM-based assertions.

### Key Changes
- **Merged Duplicate Hints**: Combined separate opening and closing tag checks (e.g. `You should have an opening <tag> tag` and `You should have a closing </tag> tag`) into single element presence checks (e.g. `You should have a <tag> element`) for `main`, `h1`, `p`, `style`, `h2`, `section`, `div`, `article`, and `footer` elements.
- **Robust DOM Assertions**: Replaced raw code regex testing with standard DOM APIs like `document.querySelector('tag')` in **Steps 1, 2, 3, 4, 5, 6, 16, 25, 42, 59, 63**.
- **Improved Traversal/Attribute checks**:
  - Refactored viewport `meta` validation in **Step 13** to query DOM attributes directly instead of matching raw code strings.
  - Refactored CSS class validation (e.g. `.established`, `.address`, `.bottom-line`) in **Steps 57, 73, 85** to check classes via the DOM instead of raw regex matching.
  - Style Isolation: Updated style tag presence checks in **Steps 6 & 11** to query style elements excluding the test framework's hidden header elements (`style:not(.fcc-hide-header)`).

These changes ensure the test runner validates the final rendered page structure instead of exact formatting/whitespaces, making the tests formatting-resilient.

### CODE REVIEW

**Review by majestic-owl448 — CHANGES_REQUESTED** [review]
the changes to `curriculum/challenges/english/blocks/learn-basic-css-by-building-a-cafe-menu/` are out of scope, the issue is asking about `workshop-cafe-menu` instead

**Comment by rohit-h11** [comment]
My bad, i interpreted the refactoring was required to be on workshop cafe menu , I will make changes to address the same. does building a cafe menu also not require similar refactoring?. should i create an issue for it.


also could u share if there is any discord or other community of freeCodeCamp contributers

**Comment by majestic-owl448** [comment]
> does building a cafe menu also not require similar refactoring?

this is part of the archived curriculum, which we are not maintaining anymore



> could u share if there is any discord or other community of freeCodeCamp contributers

forum: https://forum.freecodecamp.org
discord: https://chat.freecodecamp.org

**Review by majestic-owl448 — COMMENTED** [review]
it was not asked to remove the tests for the closing tags

for the test of the two `p` elements next to each other, does the code proposed in the issue not work?

**Review by majestic-owl448 — COMMENTED** [review]
if you force push it makes it hard to see what you changed

image

just add a new commit with your changes

**Comment by rohit-h11** [comment]
I did the proposed changes regarding closing tags and p tag adjecency issue, and added those commits here,
should i create a new PR with these proposed changes and a single commit or can this PR get reviewed ?.

**Comment by majestic-owl448** [comment]
this PR can be reviewed, but please stop force pushing, add new commits for new changes

File: curriculum/challenges/english/blocks/workshop-cafe-menu/5f356ed6cf6eab5f15f5cfe6.md [hunk 1]
```diff
@@ -13,7 +13,7 @@ Inside the opening `div` tag, add the `id` attribute with a value of `menu`.
 
 # --hints--
 
-Your opening `<div>` tag should have an `id` attribute set to `menu`.
+Your code should have a `div` element with an `id` set to `menu`.
```
Comment (majestic-owl448) [inline curriculum/challenges/english/blocks/workshop-cafe-menu/5f356ed6cf6eab5f15f5cfe6.md]: test is not changed, so now this hint change is out of scope

File: curriculum/challenges/english/blocks/workshop-cafe-menu/5f7691dafd882520797cd2f0.md [hunk 2]
```diff
@@ -14,7 +14,11 @@ Next, position the other `p` elements to be on the same line with no space betwe
 You should not have any spaces between your `p` elements.
 
 ```js
-assert.notMatch(code, /<\/p>\s+<p/);
+document.querySelectorAll('.item').forEach(item => {
```
Comment (majestic-owl448) [inline curriculum/challenges/english/blocks/workshop-cafe-menu/5f7691dafd882520797cd2f0.md:18]:
```suggestion
assert.lengthOf(document.querySelectorAll('.item'), 5);
document.querySelectorAll('.item').forEach(item => {
```

File: curriculum/challenges/english/blocks/workshop-cafe-menu/5f3cade99dda4e6071a85dfd.md [hunk 3]
```diff
@@ -11,10 +11,10 @@ You will come back to styling the menu in a few steps, but for now, go ahead and
 
 # --hints--
 
-You should have an opening `section` tag.
+You should have two `section` elements.
```
Comment (majestic-owl448) [inline curriculum/challenges/english/blocks/workshop-cafe-menu/5f3cade99dda4e6071a85dfd.md]:
```suggestion
You should add a second `section` element.
```

File: curriculum/challenges/english/blocks/workshop-cafe-menu/5f3cade99dda4e6071a85dfd.md [hunk 4]
```diff
@@ -11,10 +11,10 @@ You will come back to styling the menu in a few steps, but for now, go ahead and
...
 
 ```js
-assert.lengthOf(code.match(/<section>/ig) ,2);
+assert.lengthOf(document.querySelectorAll('section'), 2);
 ```
 
 You should have a closing `section` tag.
```
Comment (majestic-owl448) [inline curriculum/challenges/english/blocks/workshop-cafe-menu/5f3cade99dda4e6071a85dfd.md]:
```suggestion
You should have a second closing `section` tag.
```

File: curriculum/challenges/english/blocks/workshop-cafe-menu/5f344fad8bf01691e71a30eb.md [hunk 5]
```diff
@@ -11,10 +11,10 @@ Until now, you've had limited control over the presentation and appearance of yo
 
 # --hints--
 
-Your code should have an opening `<style>` tag.
+Your code should have a `style` element.
 
 ```js
-assert.match(code, /<style\s*>/i);
+assert.lengthOf(document.querySelectorAll('style:not(.fcc-hide-header)'), 1);
```
Comment (majestic-owl448) [inline curriculum/challenges/english/blocks/workshop-cafe-menu/5f344fad8bf01691e71a30eb.md]:
```suggestion
assert.exists(document.querySelector('style:not(.fcc-hide-header)'));
```

**Comment by rohit-h11** [comment]
Thanks, I have addressed the suggestions and pushed commits

File: curriculum/challenges/english/blocks/workshop-cafe-menu/5f33294a6af5e9188dbdb8f3.md [hunk 6]
```diff
@@ -14,10 +14,10 @@ Start by adding some menu content. Add a `main` element within the existing `bod
 
 # --hints--
 
-Your code should have an opening `<main>` tag.
+Your code should have a `main` element.
 
 ```js
-assert.match(code, /<main>/i);
+assert.exists(document.querySelector('main'));
```
Comment (MuhammadNiazAli) [inline curriculum/challenges/english/blocks/workshop-cafe-menu/5f33294a6af5e9188dbdb8f3.md:20]: perfected

File: curriculum/challenges/english/blocks/workshop-cafe-menu/5f332a88dc25a0fd25c7687a.md [hunk 7]
```diff
@@ -11,10 +11,10 @@ The name of the cafe is `CAMPER CAFE`. So, add an `h1` element within your `main
 
 # --hints--
 
-You should have an opening `<h1>` tag.
+You should have an `h1` element.
 
 ```js
-assert.match(code, /<h1>/i);
+assert.exists(document.querySelector('h1'));
```
Comment (MuhammadNiazAli) [inline curriculum/challenges/english/blocks/workshop-cafe-menu/5f332a88dc25a0fd25c7687a.md:17]: bug solved

File: curriculum/challenges/english/blocks/workshop-cafe-menu/5f332b23c2045fb843337579.md [hunk 8]
```diff
@@ -11,10 +11,10 @@ To let visitors know the cafe was founded in `2020`, add a `p` element below the
 
 # --hints--
 
-You should have an opening `<p>` tag.
+You should have a `p` element.
 
 ```js
-assert.match(code, /<p>/i);
+assert.exists(document.querySelector('p'));
```
Comment (MuhammadNiazAli) [inline curriculum/challenges/english/blocks/workshop-cafe-menu/5f332b23c2045fb843337579.md:17]: very nice

**Comment by MuhammadNiazAli** [comment]
i check the good this is very good work

File: curriculum/challenges/english/blocks/workshop-cafe-menu/5f3cade9fa77275d9f4efe62.md [hunk 9]
```diff
@@ -16,7 +16,9 @@ You could keep trying various percentages for the widths. Instead, use the backs
 Your `p` elements should not have any space between them.
 
 ```js
-assert.match(code, /Vanilla<\/p><p/);
+const flavor = document.querySelector('.flavor');
+const price = document.querySelector('.price');
+assert.strictEqual(flavor?.nextSibling, price);
```
Comment (majestic-owl448) [inline curriculum/challenges/english/blocks/workshop-cafe-menu/5f3cade9fa77275d9f4efe62.md:23]:
```suggestion
const flavor = document.querySelector('.flavor');
const price = document.querySelector('.price');
assert.exists(flavor);
assert.exists(price);
assert.strictEqual(flavor?.nextSibling, price);
```
if `price` does not exist anymore this passes otherwise

File: curriculum/challenges/english/blocks/workshop-cafe-menu/5f716bee5838c354c728a7c5.md [hunk 10]
```diff
@@ -72,7 +72,11 @@ assert.match(prices?.[3]?.innerText.trim(), /2\.50/);
 You should not have any spaces between your `p` elements.
 
 ```js
-assert.notMatch(code, /<\/p>\s+<p/);
+document.querySelectorAll('.item').forEach(item => {
```
Comment (majestic-owl448) [inline curriculum/challenges/english/blocks/workshop-cafe-menu/5f716bee5838c354c728a7c5.md:76]: this one needs a guard before, otherwise the test passes with an empty array resulting from `querySelectorAll`

File: curriculum/challenges/english/blocks/workshop-cafe-menu/5f7b87422a560036fd03ccff.md [hunk 11]
```diff
@@ -33,7 +33,11 @@ assert.match(lastElement?.textContent, /1\.50/);
 You should not have any spaces between your `p` elements.
 
 ```js
-assert.notMatch(code, /<\/p>\s+<p/);
+document.querySelectorAll('.item').forEach(item => {
```
Comment (majestic-owl448) [inline curriculum/challenges/english/blocks/workshop-cafe-menu/5f7b87422a560036fd03ccff.md:37]: if `querySelectorAll` finds no element, this passes

**Review by majestic-owl448 — COMMENTED** [review]
changes to `curriculum/challenges/english/blocks/workshop-cafe-menu/5f356ed6cf6eab5f15f5cfe6.md` keep being here, please remove

**Review by majestic-owl448 — CHANGES_REQUESTED** [review]
[sorry did not notice how recent the notification is]

**Comment by rohit-h11** [comment]
Thanks for the review and guidance, @majestic-owl448! I have applied the requested changes:

Reverted unrelated changes: Reverted the out-of-scope changes in 5f356ed6cf6eab5f15f5cfe6.md back to main.
Added exists checks: Applied the assert.exists(flavor) and assert.exists(price) guards in 5f3cade9fa77275d9f4efe62.md.
Added array length guards: Added assert.lengthOf(document.querySelectorAll('.item'), N) checks before the .forEach loops in 5f716bee5838c354c728a7c5.md (9 items) and 5f7b87422a560036fd03ccff.md (6 items) to prevent false positives when elements are missing.
Everything has been rebased and pushed. Please take another look when you have a moment!

File: curriculum/challenges/english/blocks/workshop-cafe-menu/5f3477ae8466a9a3d2cc953c.md [hunk 12]
```diff
@@ -14,7 +14,7 @@ Now that you have the CSS in the `styles.css` file, go ahead and remove the `sty
 You should not have any `style` tags in your code.
 
 ```js
-assert.notMatch(code, /style/i);
+assert.lengthOf(document.querySelectorAll('style:not(.fcc-hide-header)'), 0);
```
Comment (majestic-owl448) [inline curriculum/challenges/english/blocks/workshop-cafe-menu/5f3477ae8466a9a3d2cc953c.md]:
```suggestion
assert.lengthOf(document.querySelectorAll('style:not(.fcc-hide-header):not(.fcc-injected-styles)'), 0);
```

**Review by majestic-owl448 — APPROVED** [review]
Thank you for your contribution to the page! 👍
We are happy to accept these changes and look forward to future contributions. 🎉
